app/javascript/schools/settings/components/SchoolCustomize__LinkComponent.res

```diff
@@ -0,0 +1,258 @@
+
+module DestroySchoolLinkQuery = %graphql(`
+  mutation DestroySchoolLinkMutation($id: ID!) {
+    destroySchoolLink(id: $id) {
+      success
+    }
+  }
+  `)
+
+module SortSchoolLinkQuery = %graphql(`
+  mutation SortSchoolLinksMutation($id: ID!,$direction:MoveDirection!) {
+    sortSchoolLinks(id:$id,direction:$direction) {
+      links {
+        id
+        title
+        url
+        kind
+        sortIndex
+     }
+    }
+  }
+  `)
+
+let handleMoveLink = (~id, ~direction, ~dispatch, ~moveLinkCB) => {
+  dispatch(SetUpdating(true))
+  SortSchoolLinkQuery.make({id: id, direction: direction})
+  |> Js.Promise.then_(response => {
+    dispatch(SetUpdating(false))
+    open Json.Decode
+    let x = field("links", list(Customizations.decodeLink), response["sortSchoolLinks"])
+    moveLinkCB(x)
+    dispatch(SetEditing(false))
+    Js.Promise.resolve()
+  })
+  |> ignore
+  ()
+}
+
+let handleDelete = (deleting, disableDeleteCB, removeLinkCB, id, event) => {
+  event |> ReactEvent.Mouse.preventDefault
+
+  if deleting |> List.mem(id) {
+    ()
+  } else {
+    disableDeleteCB(id)
+
+    // send(SchoolLinks.DisableDelete(id))
+
+    DestroySchoolLinkQuery.make({id: id})
+    |> Js.Promise.then_(_response => {
+      removeLinkCB(id)
+      Js.Promise.resolve()
+    })
+    |> ignore
+  }
+}
+
+module UpdateSchoolLinkQuery = %graphql(`
+  mutation UpdateSchoolLinkMutation($id: ID!, $title: String, $url: String!) {
+    updateSchoolLink(id: $id, title: $title, url: $url) {
+     success
+    }
+  }
+`)
+
+let handleLinkEdit = (~dispatch, ~updateLinkCB, ~id, ~title: string, ~url) => {
+  dispatch(SetUpdating(true))
+  UpdateSchoolLinkQuery.make({id: id, title: Some(title), url: url})
+  |> Js.Promise.then_(_ => {
+    dispatch(SetUpdating(false))
+    dispatch(SetEditing(false))
+    updateLinkCB(id, title, url)
+    Js.Promise.resolve()
+  })
+  |> ignore
+  ()
+}
+
+let initialState = (title: string, url: string) => {
+  {
+    title: title,
+    url: url,
+    editing: false,
+    error: false,
+    updating: false,
+  }
+}
+
+let reducer = (state, action) => {
+  switch action {
+  | SetEditing(editing) => {...state, editing: editing}
+  | UpdateTitle(title) => {...state, title: title}
+  | UpdateUrl(url) => {...state, url: url}
+  | SetError(error) => {...state, error: error}
+  | SetUpdating(updating) => {...state, updating: updating}
+  }
+}
+
+@react.component
+let make = (
+  ~id,
+  ~title,
+  ~url,
+  ~deleting,
+  ~disableDeleteCB,
+  ~removeLinkCB,
+  ~updateLinkCB,
+  ~kind,
+  ~index,
+  ~total,
+  ~moveLinkCB,
+) => {
+  let (localState, dispatch) = React.useReducer(reducer, initialState(title, url))
+
+  <DisablingCover disabled=localState.updating message="Updating...">
+    <div
+      className={"flex justify-between items-center gap-8 bg-gray-100 text-xs text-gray-900 border rounded mt-2"}>
+      <div className="flex items-center flex-1">
+        {localState.editing
```

Comment:
It's good to extract the link editor as a separate component that handles the editor behaviour independently.

```rescript
module Editor = {
  let make = (~title, ~url, ~updateCB) {
  ... Your component logic here
  }
}
```

and use it like:

``` rescript
state.edititing ? <Editor url title updateCB/> : ...your preview logic
```